Added push-to-external application feature documentation

michizhou opened the pull request:
Added new documentation with details about usage of the push-to-external-application feature through the tools submenu and configuration of the feature that allows one to select which external application a user wishes to push to. Also modified heading to comply with markdown lint rules.

Closes #207

koppor commented:
This is a good start. Please add screenshots.

If you are good in Java-coding, I would like to take the screenshots using a TestFX test. See https://github.com/TestFX/TestFX for details on TestFX. There should be some tests in our code.

michizhou commented:
@koppor I have added some screenshots I took of the JabRef application with the push-to-external application button and dropdown menu highlighted without using TestFX. If these are not sufficient, please let me know and I can try to take the screenshots using TestFX.

koppor commented:
OK. - When reading the text, I do not see which tools are there. I remember that we have preferences where the user can configure paths etc. Is this still the case? Can you check (and provide a screenshot)?

I also found following FAQ. - Can you include the text here, too? Just the answer is enough.

#### Q: JabRef does not push to vim, although I have configured the right path and server name. What is going on?

A: You have to start vim with the option `--servername` (such as `vim --servername MyVimServer`). If you get the `Unknown option argument` message, it means your version of  vim does not include the *clientserver* feature (you can check with `vim --version`). In such a case, you have to install another version of `vim`.

michizhou commented:
@koppor If you're talking about the Preferences window that is reminiscent of the following window below, then that window will pop up when you select **Options -> Preferences** in the top toolbar.
<img width="1280" alt="Preferences_Window" src="https://user-images.githubusercontent.com/33012425/67694258-72916100-f979-11e9-8d6a-05fc7aacc425.png">
<img width="1280" alt="Preferences_Window" src="https://user-images.githubusercontent.com/33012425/67694148-4544b300-f979-11e9-8f3b-4289496be49e.png">
Should I include these screenshots in the documentation? Or did you have something else in mind?

Siedlerchr commented:
You can select the Push-App in the Preferences (External Programs) from the drop down/combobox
This then changes the icon-button in the toolbar.
See https://github.com/JabRef/jabref/issues/5375#issuecomment-547044184

I think it would be nice to illustrate it with screenshots, showing the effect and where the "Push"-Button is located.

koppor commented:
Other screenshots: https://github.com/koppor/jabref/issues/367. You can just save them and re-use them.

koppor commented:
Please also add a screenshot of TeXStudio which shows that the entry is pasted. Do you have a minimal latex document (and a minimal bibliography)?

There should be two screenshots. One for `\cite{entryOne}` and one for `\cite{entryOne,entryTwo}` so that it is illustrated, that first the complete `\cite` is inserted, then the cite is extended.

michizhou commented:
@koppor I've added additional screenshots and expanded documentation for the push-to-external application feature section as a whole. Let me know if this will suffice.

koppor commented:
Thank you for keeping working on this. Documentation writing is hard. Even for "small features"

May I ask which version you used? I remember using http://builds.jabref.org/master/. There, I did not see any drop-down menu. Did you?

Refs https://github.com/koppor/jabref/issues/367 (where I posted screenshots)

michizhou commented:
@koppor Not a problem. It's my pleasure to work on this, as I am interested in its resolution and do think it has the potential to benefit many JabRef users.

As for the version of JabRef I used, my version is 4.3.1 (the latest stable release for MacOS), and I had downloaded it from Fosshub, not http://builds.jabref.org/master/. I'll download the dev build from there (version 5.0.0) and redo the screenshots if need be.

michizhou commented:
@koppor If you could look over my most recent changes to the push application documentatioon whenever you are able, I would greatly appreciate it. Thank you.

koppor commented on en/ExternalFiles.md:
```diff
@@ -16,6 +16,32 @@ These are by default set to values that probably make sense for your operating s
+On MacOS:
+![Push to External MacOS](images/Push-External-Button-MacOS.png)
+
+On Windows:
```
Could you add two spaces (`  `) after "On Windows:"? This leads to a hard line break. (Please also at "On Mac OS:")

koppor commented on en/ExternalFiles.md:
```diff
@@ -16,6 +16,32 @@ These are by default set to values that probably make sense for your operating s
+
+When you click on the push-to-external application button, JabRef will export your selected entries to a file generated by the selected external viewer application. As an example, here is what happens when you export one entry to TexStudio.
+
+![Initial Push to External Export](images/Initial-Push-Export.png)
```
Thank you for installing the tools. The thing is, that one should have a `.tex` file opened. You can use that file: https://github.com/latextemplates/scientific-thesis-template/blob/master/latexhints-english.tex#L43

Adding a `cite` command to a BibTeX file leads to compilation errors.

![grafik](https://user-images.githubusercontent.com/1366654/68100093-55ebb080-fec6-11e9-8e72-b8184df8f065.png)

michizhou replied:
I don't quite understand what you want. When I export a bibliography in JabRef using push-to-external application, it gets all existing entries in the bibliography and I get the following: 

After exporting first entry to TexStudio
<img width="1280" alt="Screen Shot 2019-11-06 at 5 11 21 PM" src="https://user-images.githubusercontent.com/33012425/68342396-7e3b0100-00b8-11ea-99d0-f543d76de610.png">

After exporting additional (second) entry to TexStudio
<img width="1280" alt="Screen Shot 2019-11-06 at 5 14 25 PM" src="https://user-images.githubusercontent.com/33012425/68342576-e7bb0f80-00b8-11ea-8bc1-cb363e6bf77a.png">

Should I use these screenshots instead? Or are you saying that the exported file has to be in `.tex` format (which I then show screenshots for)? If it's the latter, I'd appreciate it if you could guide me on how to get the proper `.tex` file for the exported entries via JabRef (or am I supposed to use the file you linked as an example). Please let me know.

koppor commented:
Second one.

The typical workflow is: Open the bib in JabRef. manage entries there. Open
the Tex file on texstudio. Produce scientific content there. Have pdflatex
and biber do their work to produce a PDF from .tex and .bib. -- In other
words: A bib file is typically not opened in texstudio (because JabRef is
used to edit the bib file. JabRef stores the bib file directly. No need for
an export)


michizhou <notifications@github.com> schrieb am Mi., 6. Nov. 2019, 23:17:

> *@michizhou* commented on this pull request.
> ------------------------------
>
> In en/ExternalFiles.md
> <https://github.com/JabRef/help.jabref.org/pull/244#discussion_r343357811>
> :
>
> > +![Push to External MacOS](images/Push-External-Button-MacOS.png)
>
> +
>
> +On Windows:
>
> +![Push to External Windows](images/Push-External-Button-Windows.png)
>
> +
>
> +JabRef also allows you to change the external viewer application you would like to push your exports to. To do so, first go to **Options → Preferences → External programs**. Under the **Push applications** section click on the **Application to push entries to** field. This will cause a dropdown menu to appear, from which you are able to select from a list of the external viewers you have configured.
>
> +
>
> +![Select External Application](images/During-Application-Selection.png)
>
> +
>
> +Once you have made your selection and click **Save**, the push-to-external application button icon will change to match that of the selected external viewer application.
>
> +
>
> +![New Application After Select](images/After-Application-Selection.png)
>
> +
>
> +When you click on the push-to-external application button, JabRef will export your selected entries to a file generated by the selected external viewer application. As an example, here is what happens when you export one entry to TexStudio.
>
> +
>
> +![Initial Push to External Export](images/Initial-Push-Export.png)
>
>
> I don't quite understand what you want. When I export a bibliography in
> JabRef using push-to-external application, it gets all existing entries in
> the bibliography and I get the following:
>
> After exporting first entry to TexStudio
> [image: Screen Shot 2019-11-06 at 5 11 21 PM]
> <https://user-images.githubusercontent.com/33012425/68342396-7e3b0100-00b8-11ea-99d0-f543d76de610.png>
>
> After exporting additional (second) entry to TexStudio
> [image: Screen Shot 2019-11-06 at 5 14 25 PM]
> <https://user-images.githubusercontent.com/33012425/68342576-e7bb0f80-00b8-11ea-8bc1-cb363e6bf77a.png>
>
> Should I use these screenshots instead? Or are you saying that the
> exported file has to be in .tex format (which I then show screenshots
> for)? If it's the latter, I'd appreciate it if you could guide me on how to
> get the proper .tex file for the exported entries via JabRef (or am I
> supposed to use the file you linked as an example). Please let me know.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/help.jabref.org/pull/244?email_source=notifications&email_token=AAKNU7VHUPR7WNPWRM4NOEDQSM7AZA5CNFSM4JFM5OI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCKSK6VY#discussion_r343357811>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKNU7QK4TAB66CJ5XE7VZTQSM7AZANCNFSM4JFM5OIQ>
> .
>

michizhou commented:
@koppor I apologize for the relatively long time it took for me to push the latest changes, but I have replaced the erroneous TeXStudio export screenshots with proper ones this time. I also updated the descriptions related to those screenshots accordingly and took care of formatting. Please let me know if further changes are needed before final merging.

koppor commented:
Thank you for the follow up. One Nitpick: The icon in the toolbar does not match the text. The user changed to "TeXStudio". Could you provide an addtional screenshot with texstudio icon circled. Maybe first emacs (as it exists) and then texstudio (new)

![grafik](https://user-images.githubusercontent.com/1366654/68526849-62b23f00-02e0-11ea-95fa-7c7b3025af20.png)


Can you move your additoins to a new file `push-to-applications.md`? I think, the text is better placed in a separate md file than in "External files", where DOIs and external links are handled.

michizhou commented:
@koppor I replaced the incorrect Preferences window screenshot with a new one that indicates a user selection for Emacs as their new external viewer application. I also moved all my changes to a new separate file as per your suggestion. I would appreciate your feedback. Thanks!

koppor commented:
Looks good. TeXStudio should always be spelled like that in the text. Moreover, it is an "editor", not a "viewer". Can you please quickly change that?

michizhou commented:
@koppor I have made the minor edits you pointed out. Please review them when you can.

koppor commented:
Thank you for following up!

I hope, you will continue creating documentation. May I propose https://github.com/JabRef/help.jabref.org/issues/93?

michizhou commented:
@koppor It was my pleasure. I will definitely consider contributing more documentation in the future. Thank you again for all your help and support for this issue!

Siedlerchr commented:
@michizhou  Thanks again for your valuable contribution and! This is very much appreciated and will help the users of JabRef very much!

michizhou commented:
@Siedlerchr I forgot to ask, but since I have made five PRs (all of which were made in October) that have since been merged, would I be eligible for the JabRef Hacktoberfest swag? If so, I'd greatly appreciate it if you could let me know how to provide my shipping address to claim the swag as well as any other personal information I would need to provide. Thank you.

koppor commented:
@michizhou Thank you for reaching us out. Also with your help the new user documentation page at https://docs.jabref.org/ looks very nice.

We just published a blog post reflecting on #hacktoberfest. -- Find my email address on my GitHub profile. Drop me a short note with your shipping address. Please give us a few weeks to prepare the T-Shirts.

We are currently busy with preparing the 5.0 release. It would be very nice if you continued supporting us so that we can release faster!
